envoy/stats/allocator.h

```diff
@@ -58,6 +58,20 @@ class Allocator {
   virtual const SymbolTable& constSymbolTable() const PURE;
   virtual SymbolTable& symbolTable() PURE;
 
+  /**
+   * Iterate over all stats that need to be sinked. Note, that implementations can potentially hold
+   * on to a mutex that will deadlock if the passed in functors try to create or delete a stat.
+   * @param f_size functor that is provided the number of all sinked stats. Note this is called
+   *        only once.
+   * @param f_stat functor that is provided one sinked stat at a time.
+   */
+  virtual void forEachSinkedCounter(std::function<void(std::size_t)> f_size,
+                                    std::function<void(Stats::Counter&)> f_stat) PURE;
+  virtual void forEachSinkedGauge(std::function<void(std::size_t)> f_size,
+                                  std::function<void(Stats::Gauge&)> f_stat) PURE;
+  virtual void forEachSinkedTextReadout(std::function<void(std::size_t)> f_size,
+                                        std::function<void(Stats::TextReadout&)> f_stat) PURE;
```

Comment:
it might be convenient for callers to receive these as `const Stats::CounterSharedPtr&` etc rather than `Stats::Counter&`. That way if a caller decided it wanted to collect them and sort them, it could. I admit that would defeat the purpose of the new API, but it would also mean you could implement the old API on top of the new one, and remove the old API (potentially).

It would also make it easier to (for example) have an admin option to do some alternative sorting, or doing a custom sort after doing its own regex filtering.

And I don't think you'd lose any performance or generality.

Reply:
That would require creating the shared pointer from raw pointers and passing it to the functor. Would it not be better for the functor to do that (and pay the cost) if it needs to?

Reply:
Yes you are right. I was thinking you had the RefcountPtr (not std::shared_ptr) held in the set you were iterating over, but actually it's just got raw pointers.

And because these are holding the ref-count in a member variable you can actually construct RefcountPtr from a raw ptr if needed:

https://github.com/envoyproxy/envoy/blob/74d5ef8bb36bfcbdd89555a4ea020519033699f5/envoy/stats/refcount_ptr.h#L35